test/GPv2Settlement/Settle/Settle.t.sol

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.8.26;
+
+import {Helper} from "../Helper.sol";
+
+import {GPv2Settlement, GPv2Interaction} from "src/contracts/GPv2Settlement.sol";
+
+import {SettlementEncoder} from "test/libraries/encoders/SettlementEncoder.sol";
+
+// solhint-disable func-name-mixedcase
+contract Settle is Helper {
+    using SettlementEncoder for SettlementEncoder.State;
+
+    function test_allowlist_rejects_transactions_from_non_solvers() public {
+        vm.expectRevert("GPv2: not a solver");
+        settle(encoder.encode(settlement));
+    }
+
+    function test_allowlist_accepts_transactions_from_solvers() public {
+        vm.prank(solver);
+        settle(encoder.encode(settlement));
+    }
+
+    function test_emits_a_settlement_event() public {
+        vm.expectEmit(address(settlement));
+        emit GPv2Settlement.Settlement(solver);
+        vm.prank(solver);
+        settle(encoder.encode(settlement));
+    }
+
+    function test_reverts_if_encoded_interactions_has_incorrect_number_of_stages() public {
+        for (uint256 i = 1; i < 3; i++) {
+            GPv2Interaction.Data[][] memory interactions = new GPv2Interaction.Data[][](i * 2);
+            assertTrue(interactions.length != 3, "incorrect interaction array length test setup");
+            vm.expectRevert();
+            // test requires malformed interactions array, therefore use encodeWithSelector
+            (bool revertsAsExpected,) = address(settlement).call(
+                abi.encodeWithSelector(
+                    GPv2Settlement.settle.selector, new bytes32[](0), new uint256[](0), new bytes[](0), interactions
+                )
```

Comment:
An issue I see with this approach is that the for loop is testing the same thing in both iterations, that is, the dynamic array `interaction` doesn't encode into something that can be parsed by the `settle` function; it should also fail when `interactions.length == 3`.
To be fair it's unclear what the original test is doing, but considering that the call wouldn't revert with three elements I'd say that this can't be done in a loop and we should manually define two things: `Interactions[][2] interactions` and `Interactions[][4] interactions`. To be thorough, we could additionally test that `Interactions[][] memory interactions = new GPv2Interaction.Data[][](3)` causes a revert.
There is an argument for removing this test, since this is less meaningful when the test itself is done in Solidity,

Reply:
Agreed, given that the tests are now in solidity, type checking is enforced, so will remove this test.